driver-core/src/main/com/mongodb/ConnectionString.java

```diff
@@ -296,6 +297,21 @@ public class ConnectionString {
      * @since 3.0
      */
     public ConnectionString(final String connectionString) {
+        this(connectionString, null);
+    }
+
+    /**
+     * Creates a ConnectionString from the given string with the given {@link DnsClient}.
+     *
+     * <p>If setting {@link MongoClientSettings#getDnsClient()} explicitly, care should be taken to call this constructor with the same
```

Comment:
This is unfortunate but I don't see a way around it, as ConnectionString is responsible for TXT record resolution.

Reply:
I just can't think of a concrete plan for it, even if we allow breaking changes.  What happens if you call any of the `applyConnectioString` methods on the setting classes, and then pass a MongoClientSettings in to `MongoClient#create`.  If you then try to apply the TXT record to the `MongoClientSettings` you have the problem that the settings are supposed to trump the TXT record and so you have to remember in the settings the difference between a value that is the default and the same value that was specified explicitly.  Tricky.

Reply:
# On getters

A `ConnectionString` that does not do what it's not supposed to do should verify the format of the string, may destructure it on components, but should not provide any getters to users (even if currently implementing most of them don't require any IO, who knows what may happen in the future). Therefore, a `MongoClientSettings`/`MongoClientSettings.Builder` that was fed a `ConnectionString` should not provide any getters to users either. I don't think these getters are actually required by users: even if in some bizarre cases a user want to know what's inside a `MongoClientSettings`/`Builder`, he may pack that data in any form needed together with `MongoClientSettings`/`Builder` in another structure (a weird requirement needs some additional work on the user side, and is not supported out of the box, that's fine).

If for some reason we end up strongly wanting to provide getters (I hope we don't), we may have two sets of `MongoClientSettings`/`Builder` types: as long as a `ConnectionString` (or any other future mechanism that requires IO or customizable logic similarly to the `+srv` connection string modifier) is not used, we use the types with getters, otherwise we return a type without getters (see the next section for more details).

# On the path forward

I see two paths here (only the first one seems practical, I don't even know why I describe the second one):

1. We allow applying a connection string to settings only when a new settings object is created. For example, this means that `Builder.applyConnectionString` should create a new `Builder`, instead of modifying `this`. If we have two sets of `MongoClientSettings`/`Builder` types (see the previous section for more details), then `Builder.applyConnectionString` returns a `Builder` without getters.
2. We maintain a sequence of updates in `MongoClientSettings`/`Builder`:
  - `applyConnectionString` introduces a new element to the sequence;
  - any other setter is merged together with the immediate previous update as long as that update is not `applyConnectionString`, otherwise it also introduces a new element to the sequence.
At the point when we are ready to resolve `ConnectionString` (and any other settings that require IO or customizable logic), we resolve and merge the updates in the sequence.

Reply:
Does the above seem like something we could explore further when designing the 5.0 API improvements? If yes, I'll create a ticket. If not, or if we don't really want to really have any significant API changes in 5.0, then I won't.

I went to such length in this discussion because the requirement `If setting {@link MongoClientSettings#getDnsClient()} explicitly, care should be taken to call this constructor with the same {@link DnsClient}` screams at me that we failed at the API design here.

Reply:
I found something related in the SDAM spec https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring-summary.rst:

_"A multi-threaded or asynchronous client's constructor MUST NOT do any I/O. ... The justification is that if clients can be constructed when the deployment is in some states but not in other states, it leads to an unfortunate scenario: When the deployment is passing through a strange state, long-running applications may keep working, but any applications restarted during this period fail."_

Thile the SDAM spec focuses on specifically client constructor and the state of the MongoDB deployment, the principle of not doing IO in constructors should always be followed for the same reason pointed out in the spec: object creation should not fail because of the temporary unfortunate external state. The constructor of `ConnectionString` may fail due not network issues, which may result in the whole application failing to start. Despite the constructor of a `ConnectionString` formally not being part of the client constructor in the Java driver, I think, in essence the our driver violates the idea that the SDAM spec expresses.

Reply:
Maybe what was bad was supporting TXT records. :).  I can only say that we knew at the time we were violating the SDAM spec recommendation but didn't see a lot of good options given the API we had to support at the time.

Reply:
Created https://jira.mongodb.org/browse/JAVA-4944.